include/souper/Infer/InstSynthesis.h

```diff
@@ -0,0 +1,300 @@
+// Copyright 2014 The Souper Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOUPER_INST_SYNTHESIS_H
+#define SOUPER_INST_SYNTHESIS_H
+
+#include "souper/Inst/Inst.h"
+#include "souper/SMTLIB2/Solver.h"
+
+#include <map>
+#include <set>
+#include <vector>
+#include <utility>
+
+/*
+  This class is based on the "Synthesis of Loop-free Programs" approach
+  by Gulwani et al. presented at PLDI'11.
+  Paper:  http://dl.acm.org/citation.cfm?id=1993506
+  Slides: http://www.eecs.berkeley.edu/~jha/pres/pldi11.pdf
+
+  The implementation is derived from the description in the paper, but it
+  also contains additional steps that were necessary to apply this synthesis
+  method in Souper.
+*/
+
+namespace souper {
+
+/// A location variable identifies an input variable, the output variable, or
+/// input/output of a library component. The encoding has the following
+/// semantics: If the first integer is 0, it's an input var or a constant
+/// where the second integer describes its number. If the second integer is 0,
+/// it's an output var and the first integer describes the number of a
+/// component. The output variable is encoded by second integer being 0 and
+/// the first integer being equal to the number of library components plus one.
+/// Otherwise, it's a component input where the first integer describes the
+/// number of a component and the second integer describes it's input number.
+/// Note that constant components (no inputs, one output) are not treated as
+/// ordinary library components, but rather as inputs.
+/// For example:
+/// 0_1, 0_2, 0_3 describe the input variables 1, 2, and 3.
+/// 1_0, 2_0, 3_0 describe the outputs of components 1, 2, and 3.
+/// 1_1, 2_2, 3_1 describe the first input of component one, the second input of
+/// component two, and the first input of component three.
+/// 4_0 describes the final output (assuming there are only three components)
+typedef std::pair<unsigned, unsigned> LocVar;
+
+/// A mapping from a location variable to it's concrete value (Inst::Const).
```

Comment:
Argh, @jeroenk pointed to a number of these already. I thought I got rid of these, but obviously not.